### ISSUE TITLE
Bogus NPE if any() is used with primitive type arguments

### ISSUE THREAD

**Issue by axos88** [mockito/mockito#822]
Mockito: 2.3.0, Java 8

When mocking a method invocation, if any() is used instead of anyInt() for exampe in the .when method call, a bogus NPE is thrown. There is quite a lot of explanations to the root cause of that, but it does not contain this hint.

I would expect either a compilation error or a more explanatory exception.
Code: https://gist.github.com/axos88/163061ed82fcfdd9756d60d12eba8809

```
java.lang.NullPointerException
	at xxx.WtfTest.testCaseNPE(WtfTest.java:34)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.mockito.internal.runners.SilentJUnitRunner.run(SilentJUnitRunner.java:39)
	at org.mockito.internal.runners.StrictRunner.run(StrictRunner.java:39)
	at org.mockito.junit.MockitoJUnitRunner.run(MockitoJUnitRunner.java:103)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)


You cannot use argument matchers outside of verification or stubbing.
Examples of correct usage of argument matchers:
    when(mock.get(anyInt())).thenReturn(null);
    doThrow(new RuntimeException()).when(mock).someVoidMethod(anyObject());
    verify(mock).someMethod(contains("foo"))

Also, this error might show up because you use argument matchers with methods that cannot be mocked.
Following methods *cannot* be stubbed/verified: final/private/equals()/hashCode().
Mocking methods declared on non-public parent classes is not supported.

```

**Comment by axos88** [comment]
There is also some weirdness that the mockFoo's value is set to null between the invocation of when() and .foo(), on line 33.

**Comment by ChristianSchwarz** [comment]
@axos88 
This behaviour can't be avoided cause java doesn't allow return type overloads like `int any()`, `T any()` , ...,  therefore Mockito provides any<primitve>() methods.

**Comment by bric3** [comment]
**tl;dr**

This cannot be avoided as @ChristianSchwarz is writing. That's one of corner area of Java itself where can't really do something. 

**long story**

With the signature of `any()`, `javac` thinks 

* that `any()` can return an `Integer` => insert cast Object to Integer
* that the `Integer` can be unboxed => invoke `intValue`

However `any`

* has return type `T` whose implicit upper bound is `Object` => `Integer` matches for `javac`
* returns `null` reference => OK for `Integer`

The NPE is raised on the callsite by code that wasn't explicitly written, and we can't really do something here. Note the javadoc of `any()` do say to use the primitive variants :

```
For primitive types use {@link #anyChar()} family...
```

**Comment by axos88** [comment]
I understand, but a not in the output of the exception about this possibility would be welcome:

```
You cannot use argument matchers outside of verification or stubbing.
Examples of correct usage of argument matchers:
    when(mock.get(anyInt())).thenReturn(null);
    doThrow(new RuntimeException()).when(mock).someVoidMethod(anyObject());
    verify(mock).someMethod(contains("foo"))

Also, this error might show up because you use argument matchers with methods that cannot be mocked.
Following methods *cannot* be stubbed/verified: final/private/equals()/hashCode().
Mocking methods declared on non-public parent classes is not supported.
```

**Comment by bric3** [comment]
I'm not sure this is related, what fails is a NullPointerException that is raised by the JVM on the callsite of the matcher.
The mockito exception message seems related to another issue, and the exception type that have this message can only be `InvalidUseOfMatchersException`.

**Comment by axos88** [comment]
It seems like the NPE has an InvalidUseOFMatchersException as its cause

**Comment by bric3** [comment]
Ah I see the issue, JUnit Runner performs an additional check after a test, and then report a matcher misuse. Because the matchers were not _consumed_ byt the mock, the NPE is raised first.

Indeed we can mention that in the message.
Thanks for insisting on this !

**Comment by bric3** [comment]
Closing this one since there's now the PR #830

**Comment by jamesaspence** [comment]
Hi @bric3 , I believe this issue is appearing again, please see my below junit example:
```
public class MockitoTest {

    @Test
    public void testDoAThing() {
        final TestThing thing = Mockito.mock(TestThing.class);

        Mockito.when(thing.doAThing(Mockito.any(), Mockito.any()))
            .thenReturn("asd");
    }

    public static class TestThing {
        public String doAThing(String foo, boolean second) {
            return second ? foo + "asd" : foo;
        }
    }
}
```

When running this test I get the following output:
```
java.lang.NullPointerException
	at com.mytest.MockitoTest.testDoAThing(MockitoTest.java:12)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

No mention of the above error or any real indication what is happening - I ended up using trial and error to determine what was going wrong.
